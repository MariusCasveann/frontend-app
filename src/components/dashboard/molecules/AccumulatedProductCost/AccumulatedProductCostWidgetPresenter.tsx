import React from 'react';
import * as chartjs from 'chart.js';
import { Doughnut } from 'react-chartjs-2';
import LoadingOrErrorPresenter from '../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter';
import { hasChartData, hasNoChartData } from '../../helpers/checkChartData';
import '../common/css/CostWidget.css';
import { CardLegend } from '../../../common/molecules/CardLegend/CardLegend';
import { colors } from '../../../../config/colors';
import { MESSAGE_WHEN_NO_DATA } from '../../../../utils/constants';

interface AccumulatedProductCostProps {
    formattedLabels: string[];
    loading: boolean;
    error: string;
    chartData?: chartjs.ChartData;
}

export default ({ chartData, formattedLabels, loading, error }: AccumulatedProductCostProps) => {
    if (loading || error) {
        return <LoadingOrErrorPresenter error={error} loading={loading} />;
    }
    if (!loading && hasChartData(chartData)) {
        return (
            <div className="domain-legend-container">
                {chartData && <CardLegend labels={formattedLabels} colors={colors} />}
                {chartData && (
                    <div className="canvas-container">
                        <Doughnut
                            data={chartData}
                            options={{
                                responsive: true,
                                legend: { display: false }
                            }}
                        />
                    </div>
                )}
            </div>
        );
    } else if (hasNoChartData(chartData)) {
        return <p>{MESSAGE_WHEN_NO_DATA}</p>;
    }

    return null;
};
