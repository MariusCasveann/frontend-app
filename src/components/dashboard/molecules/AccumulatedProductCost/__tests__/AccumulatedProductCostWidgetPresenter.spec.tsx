import * as React from 'react';
import renderer from 'react-test-renderer';
jest.mock('react-chartjs-2', () => ({ Doughnut: 'Doughnut' }));
jest.mock('../../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter', () => 'LoadingOrErrorPresenter');
jest.mock('../../../../common/molecules/CardLegend/CardLegend', () => ({ CardLegend: 'CardLegend' }));
import AccumulatedProductCostWidgetPresenter from '../AccumulatedProductCostWidgetPresenter';

const mockData = {
    labels: [
        'Cloud Build(424.15)',
        'Source Repository(1.5)',
        'Cloud Key Management Service (KMS)(168.46)',
        'Compute Engine(9261.77)'
    ],
    datasets: [
        {
            data: [424.15, 1.5, 168.46, 9261.77],
            backgroundColor: ['rgb(109,46,235)', 'rgb(222,28,70)', 'rgb(220,159,172)', 'rgb(125,62,251)']
        }
    ]
};

const mockEmptyData = {
    labels: [],
    datasets: [
        {
            data: [],
            backgroundColor: [],
            label: 'Accumulated product cost'
        }
    ]
};

const mockFormattedLabels = [
    'BigQuery (16958.16 EUR)',
    'Cloud Build (10.53 EUR)',
    'Cloud DNS (0.56 EUR)',
    'Cloud Functions (0 EUR)',
    'Cloud Key Management Service (KMS) (1.11 EUR)'
];

describe('AccumulatedProductCostWidgetPresenter', () => {
    describe('when there are chart data', () => {
        it('renders a Doughnut', () => {
            const wrapper = renderer
                .create(
                    <AccumulatedProductCostWidgetPresenter
                        formattedLabels={mockFormattedLabels}
                        error={''}
                        loading={false}
                        chartData={mockData}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when there are no chart data', () => {
        it('doesnt render a Doughnut', () => {
            const wrapper = renderer
                .create(
                    <AccumulatedProductCostWidgetPresenter
                        formattedLabels={mockFormattedLabels}
                        error={''}
                        loading={false}
                        chartData={mockEmptyData}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is undefined', () => {
        it('renders loading', () => {
            const wrapper = renderer
                .create(
                    <AccumulatedProductCostWidgetPresenter
                        formattedLabels={mockFormattedLabels}
                        error={''}
                        loading={true}
                        chartData={undefined}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is undefined', () => {
        it('renders an error', () => {
            const wrapper = renderer
                .create(
                    <AccumulatedProductCostWidgetPresenter
                        formattedLabels={mockFormattedLabels}
                        error={'Error'}
                        loading={false}
                        chartData={undefined}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
