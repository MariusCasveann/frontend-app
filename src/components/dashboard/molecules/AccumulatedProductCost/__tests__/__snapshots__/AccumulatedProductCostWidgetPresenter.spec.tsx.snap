// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccumulatedProductCostWidgetPresenter when chart data is undefined renders an error 1`] = `
<LoadingOrErrorPresenter
  error="Error"
  loading={false}
/>
`;

exports[`AccumulatedProductCostWidgetPresenter when chart data is undefined renders loading 1`] = `
<LoadingOrErrorPresenter
  error=""
  loading={true}
/>
`;

exports[`AccumulatedProductCostWidgetPresenter when there are chart data renders a Doughnut 1`] = `
<div
  className="domain-legend-container"
>
  <CardLegend
    colors={
      Array [
        "#0c3d8a",
        "#859ec4",
        "#D8DADA",
        "#487ebd",
        "#007898",
        "#7fbbcb",
        "#58737a",
        "#92aab1",
        "#eee7a5",
        "#f1e364",
        "#f8b100",
        "#dcaeb9",
        "#ad405a",
        "#d9d4b9",
        "#a19351",
        "#886b22",
        "#7d7a58",
        "#564e02",
        "#c1dfc4",
        "#7dab6e",
        "#008d3a",
        "#36542c",
        "#263428",
      ]
    }
    labels={
      Array [
        "BigQuery (16958.16 EUR)",
        "Cloud Build (10.53 EUR)",
        "Cloud DNS (0.56 EUR)",
        "Cloud Functions (0 EUR)",
        "Cloud Key Management Service (KMS) (1.11 EUR)",
      ]
    }
  />
  <div
    className="canvas-container"
  >
    <Doughnut
      data={
        Object {
          "datasets": Array [
            Object {
              "backgroundColor": Array [
                "rgb(109,46,235)",
                "rgb(222,28,70)",
                "rgb(220,159,172)",
                "rgb(125,62,251)",
              ],
              "data": Array [
                424.15,
                1.5,
                168.46,
                9261.77,
              ],
            },
          ],
          "labels": Array [
            "Cloud Build(424.15)",
            "Source Repository(1.5)",
            "Cloud Key Management Service (KMS)(168.46)",
            "Compute Engine(9261.77)",
          ],
        }
      }
      options={
        Object {
          "legend": Object {
            "display": false,
          },
          "responsive": true,
        }
      }
    />
  </div>
</div>
`;

exports[`AccumulatedProductCostWidgetPresenter when there are no chart data doesnt render a Doughnut 1`] = `
<p>
  No data available, please select another time and/or domain/product.
</p>
`;
